src/data_info/services/rename.clj

```diff
@@ -22,42 +22,55 @@
-        extract-paths (fn [{:keys [data type]}]
-                        (condp = type
-                          "data-move"         (concat [(:destination data)] (:sources data))
-                          "data-rename"       (map #(get data %) [:destination :source])
-                          "data-delete"       (concat (:paths data) (vals (:trash-paths data)))
-                          "data-delete-trash" (:trash-paths data)
-                          "data-restore"      (concat (:paths data)
-                                                      (map :restored-path (vals (:restoration-paths data))))
-                          nil)) ;; ugh, gross
-        mapcat-paths (mapcat extract-paths eligible-tasks)
-        locked-paths (reduce conj #{} mapcat-paths)
-        path-matches (fn [path] (or
-                                  (get locked-paths path)
-                                  (some #(string/starts-with? (str % "/") path)
-                                        locked-paths)))
-        matching-paths (filterv path-matches paths)]
-    (if (seq matching-paths)
-      (throw+ {:error_code error/ERR_CONFLICT
-               :paths      matching-paths}))))
+  ([sources dests]
+   (validate-unlocked (concat sources dests)))
+  ([paths]
+   ;; We need to:
+   ;; Fetch all async tasks that could cause it to be locked (data-move, data-rename, data-delete, data-delete-trash, data-restore)
+   ;; only include non-completed ones, i.e. EndDateSince sometime in the future + include null end
+   ;; Pull out their paths.
+   ;; data-move: sources, destination
+   ;; data-rename: source, destination
+   ;; data-delete: paths, trash-paths (not that the latter is real likely)
+   ;; data-delete-trash: trash-paths
+   ;; data-restore: paths, restoration-paths
+   ;; Then we need to check the paths. No source or destination can be the source or destination of another move.
+   ;; This includes exact matches but also prefix matches in both directions:
+   ;; If a locked source plus '/' is a prefix of a new source, we're moving the new source with both processes
+   ;; If a locked destination plus '/' is a prefix of a new source, we're trying to move the new source out while it's still being moved in
+   ;; If a new source plus '/' is a prefix of a locked source, we're moving the locked source with both processes
+   ;; If a new source plus '/' is a prefix of a locked destination, we're  trying to move something in while part of it (the locked dest) is being moved out
+   ;; Locked source + '/' prefix of new dest, we're moving new dest in while the parent is being moved out
+   ;; Locked dest + '/' prefix of new dest, we're moving stuff into new dest with two processes
+   ;; New dest + '/' prefix of locked source, we might be moving stuff into locked source while it's being moved out
+   ;; New dest + '/' prefix of locked dest, we're moving stuff into locked dest with two processes
```

Comment:
I'm still not sure about these 2 cases.

> New dest + '/' prefix of locked source

Could this happen if I'm moving `~/A/B.txt` somewhere else, then I'm trying to move some other `C.txt` into `~/A/` (`~/A/` is new dest and a prefix of locked `~/A/B.txt`)?
Shouldn't this be allowed?


> New dest + '/' prefix of locked dest

Could this happen if I'm moving `~/D.txt` into `~/A/B/C/`, then I'm trying to move some other `~/E.txt` into `~/A/B/` (`~/A/B/` is new dest and a prefix of locked `~/A/B/C/`)?
Shouldn't this be allowed?

Reply:
> New dest + '/' prefix of locked dest

Or is a locked dest `~/A/B/C/` if I'm moving `~/C/` into `~/A/B/`, and then if I try to move some other `~/E.txt` into `~/A/B/` (`~/A/B/` is new dest and a prefix of locked `~/A/B/C/`)?
This should still be allowed?

Reply:
So for all the prefix-related stuff, it only matters if whatever is being used for the prefix is a folder, first of all (for files, only the exact matches can apply).

For `new dest + '/' prefix of locked source`, the case we're looking at is:
 * There's a move already in progress that is moving `~/A/B/C` to wherever (let's call it move 1), which will make `~/A/B/C` a locked source path.
 * The move currently being requested (move 2) is moving something called `B` into `~/A/`, making `~/A/B` the new destination path
 * `~/A/B/` is a prefix of `~/A/B/C`, so this move should not be allowed.
 
 For `new dest + '/' prefix of locked dest`, the case is:
  * There's a move already in progress that is moving something called `C` into `~/A/B`, which would make `~/A/B/C` a locked destination path.
  * The move currently being requested is the same as move 2 above: `B` into `~/A/`, so `~/A/B` is the new destination path.
  * Once again, `~/A/B/` is a prefix of `~/A/B/C` and should not be allowed.
  
As you may note, both of these should actually be prevented other ways too, because we're also checking that the destination paths don't exist when we initiate a move. But IRODS isn't great about being completely transactional, so I think it's not impossible that there might be a window where the move is in progress but `~/A/B` for some reason isn't showing up as still existing. Both cases aren't things that should be allowed, I think, and importantly this means that we don't have to handle destinations and sources differently, since prefixes either direction with either thing are all bad.

Just to run through your example cases:

* `~/A/B.txt` being moved somewhere else (`/elsewhere/`), the locked paths will be `~/A/B.txt` and `/elsewhere/B.txt` (due to the `add-destination-to-basenames` change, this is how it will work with everything -- the `data-move` type is the only one that listed a destination folder instead of individual destinations). If you then try to move `/wherever/C.txt` to `~/A/C.txt`, those two paths will be the new source and new dest being checked, and neither will match anything with B.txt in it.
* `~/D.txt` into `~/A/B/C/` will lock `~/D.txt` and `~/A/B/C/D.txt`. Again due to `add-destination-to-basenames`/pre-existing behavior in the other types, nothing including `E.txt` will match as a prefix either direction with these.
* `~/C/` into `~/A/B/` will lock `~/C` and `~/A/B/C` (the latter because of `add-destination-to-basenames`), so once again it won't fail with anything `E.txt`

Mostly I think the `add-destination-to-basenames` change might be the missing link. A "move" with a destination of `/whatever/folder/name` won't directly lock `/whatever/folder/name`, it will lock each actual destination _within_ `/whatever/folder/name/`

Reply:
One thing that's just occurred to me, though, is that I should probably be removing all trailing slashes from things being used for prefix checks, then adding exactly one. That way if an async task has a trailing slash in it in the DB, we won't be trying to prefix match `/whatever/folder//` against anything (which it probably won't ever match)

Reply:
OK, this makes sense, and I think it was how I envisioned it originally, but after reviewing the code and looking at the destination in `add-destination-to-basenames`, I think I tripped myself up and thought that destination was the same as "new dest" in the comments. So this explanation clears that up 👍 

> One thing that's just occurred to me, though, is that I should probably be removing all trailing slashes from things being used for prefix checks, then adding exactly one. That way if an async task has a trailing slash in it in the DB, we won't be trying to prefix match `/whatever/folder//` against anything (which it probably won't ever match)

I think I've seen a similar function already in our data-info or lib code somewhere, that ensures a final `/` by either appending the `/`, or returning the original path if it already `ends-with?` a `/`.

Reply:
Yeah, you're right. It's even already used in the same namespace, haha. I changed it to use that. It could still theoretically be a problem if somehow something ends up _in the database_ with a double slash, but hopefully we aren't doing that to ourselves so I think I won't worry about it, lol.